File: engine/openscript/library/transcendental.py
"""Integer interval arithmetic for the elementary recipe in section 20.10.2."""
import math
import struct

DEFAULT_CONSTANTS = {}
MASK = (1 << 52) - 1

def up(n, d):
    return -((-n) // d)

def parts(x):
    b = int.from_bytes(struct.pack('>d', abs(x)), 'big')
    e = b >> 52
    return (b if e == 0 else (b & MASK) | (1 << 52), -1074 if e == 0 else e - 1075)

def fixed(x, p):
    m, e = parts(x)
    shift = e + p
    return (m << shift, m << shift) if shift >= 0 else (m >> -shift, up(m, 1 << -shift))

def round_dyadic(n, e):
    if not n:
        return 0.0
    negative = n < 0
    n = abs(n)
    q = max(-1074, n.bit_length() - 1 + e - 52)
    shift = q - e
    if shift <= 0:
        m = n << -shift
    else:
        d = 1 << shift
        m, r = divmod(n, d)
        if r * 2 > d or (r * 2 == d and m & 1):
            m += 1
    if not m:
        return 0.0
    if m >= 1 << 53:
        m >>= 1
        q += 1
    if q > 971:
        return None
    b = m if m < 1 << 52 else ((q + 1075) << 52) | (m & MASK)
    if negative:
        b |= 1 << 63
    return struct.unpack('>d', b.to_bytes(8, 'big'))[0]

def atanh(a, b, p):
    if not a:
        return 0, 0
    s = 1 << p
    zl, zh = a*s//b, up(a*s, b)
    sl, sh = zl*zl//s, up(zh*zh, s)
    pl, ph, lo, hi, odd = zl, zh, 0, 0, 1
    while True:
        lo += pl // odd
        hi += up(ph, odd)
        pl, ph, odd = pl*sl//s, up(ph*sh, s), odd+2
        if ph <= 2:
            return 2*lo, 2*(hi+up(9*ph, 8*odd))

def ln2(p):
    if p != 160:
        return atanh(1, 3, p)
    if 'two' not in DEFAULT_CONSTANTS:
        DEFAULT_CONSTANTS['two'] = atanh(1, 3, p)
    return DEFAULT_CONSTANTS['two']

def log_interval(x, p):
    m, e = parts(x)
    d, k = 1 << (m.bit_length()-1), e+m.bit_length()-1
    lo, hi = atanh(m-d, m+d, p)
    cl, ch = ln2(p)
    return (lo+k*cl, hi+k*ch) if k >= 0 else (lo+k*ch, hi+k*cl)

def ln10(p):
    if p != 160:
        return log_interval(10,p)
    if 'ten' not in DEFAULT_CONSTANTS:
        DEFAULT_CONSTANTS['ten'] = log_interval(10,p)
    return DEFAULT_CONSTANTS['ten']

def quotient(n, d, p):
    pairs = [(a*(1 << p), b) for a in n for b in d]
    return min(a//b for a,b in pairs), max(up(a,b) for a,b in pairs)

def exp_range(xl, xh, positive, p):
    s = 1 << p
    cl, ch = ln2(p)
    k = xl // ch
    rl, rh = xl-k*ch, xh-k*cl
    if rl < 0 or rh >= s:
        return None
    tl = th = lo = hi = s
    n = 0
    while True:
        n += 1
        tl, th = tl*rl//(s*n), up(th*rh,s*n)
        if th <= 1:
            hi += 2*th
            break
        lo += tl
        hi += th
    return (lo,hi,k-p) if positive else (s*s//hi,up(s*s,lo),-k-p)

def exp_interval(x, p):
    xl, xh = fixed(x, p)
    return exp_range(xl, xh, x >= 0, p)

def exp_endpoint(integer, p):
    """Enclose exp(integer / 2**p), with no binary64 intermediate."""
    cutoff = 1024 << p
    if integer >= cutoff:
        return None, None
    if integer <= -cutoff:
        return 0.0, 0.0
    if integer == 0:
        return 1.0, 1.0
    magnitude = abs(integer)
    interval = exp_range(magnitude, magnitude, integer > 0, p)
    if interval is None:
        return None
    lo, hi, exponent = interval
    return round_dyadic(lo, exponent), round_dyadic(hi, exponent)

def transcendental(kind, x, start_precision=160):
    if x is None or not math.isfinite(x) or (kind != 'exp' and x <= 0):
        return None
    if kind == 'exp':
        if x >= 1024:
            return None
        if x <= -1024:
            return 0.0
        if abs(x) < 2**-60:
            return 1.0
    else:
        if x == 1:
            return 0.0
        if kind == 'log2':
            m,e = parts(x)
            if not m & (m-1):
                return float(e+m.bit_length()-1)
        if kind == 'log10':
            m,e = parts(x)
            integer = m << e if e >= 0 else m >> -e if m % (1 << -e) == 0 else 0
            count = 0
            while integer > 1 and integer % 10 == 0:
                integer //= 10
                count += 1
            if integer == 1:
                return float(count)
    p = max(64, start_precision)
    while True:
        if kind == 'exp':
            interval = exp_interval(x,p)
            if interval is None:
                p += 80
                continue
            lo,hi,e = interval
        else:
            raw = log_interval(x,p)
            lo,hi = raw if kind == 'log' else quotient(raw,ln2(p) if kind == 'log2' else ln10(p),p)
            e = -p
        left,right = round_dyadic(lo,e),round_dyadic(hi,e)
        if left == right:
            return left
        p += 80

File: src/core/stdlib/maths/transcendental.ts
/** Exact integer enclosures for section 20.10.2. No host elementary calls. */
export type ElementaryKind = 'exp' | 'log' | 'log10' | 'log2';
export type Interval = readonly [bigint, bigint];
const view = new DataView(new ArrayBuffer(8));
const mask = (1n << 52n) - 1n;
let defaultTwo: Interval | undefined;
let defaultTen: Interval | undefined;

export const bitLength = (x: bigint): number => x === 0n ? 0 : x.toString(2).length;
export const floor = (n: bigint, d: bigint): bigint => n >= 0n ? n / d : -((-n + d - 1n) / d);
export const ceil = (n: bigint, d: bigint): bigint => -floor(-n, d);

export function parts(x: number): readonly [bigint, number] {
  view.setFloat64(0, Math.abs(x));
  const bits = view.getBigUint64(0), exponent = Number(bits >> 52n);
  return [exponent === 0 ? bits : (bits & mask) | (1n << 52n), exponent === 0 ? -1074 : exponent - 1075];
}

function fixed(x: number, precision: number): Interval {
  const [m, e] = parts(x), shift = e + precision;
  return shift >= 0 ? [m << BigInt(shift), m << BigInt(shift)]
    : [m >> BigInt(-shift), ceil(m, 1n << BigInt(-shift))];
}

/** Round an exact dyadic endpoint directly, with the subnormal spacing floor. */
export function rounded(numerator: bigint, exponent: number): number | null {
  if (numerator === 0n) return 0;
  const negative = numerator < 0n;
  const magnitude = negative ? -numerator : numerator;
  let spacing = Math.max(-1074, bitLength(magnitude) - 1 + exponent - 52);
  const shift = spacing - exponent;
  let significand: bigint;
  if (shift <= 0) significand = magnitude << BigInt(-shift);
  else {
    const divisor = 1n << BigInt(shift), remainder = magnitude % divisor;
    significand = magnitude / divisor;
    if (remainder * 2n > divisor || (remainder * 2n === divisor && (significand & 1n))) significand++;
  }
  if (significand === 0n) return 0;
  if (significand >= 1n << 53n) { significand >>= 1n; spacing++; }
  if (spacing > 971) return null;
  const bits = significand < 1n << 52n ? significand
    : (BigInt(spacing + 1075) << 52n) | (significand & mask);
  view.setBigUint64(0, bits | (negative ? 1n << 63n : 0n));
  return view.getFloat64(0);
}

/** Twice the positive odd-power sum, for 0 <= a/b <= 1/3. */
function oddSeries(a: bigint, b: bigint, precision: number): Interval {
  if (a === 0n) return [0n, 0n];
  const unit = 1n << BigInt(precision), low = a * unit / b, high = ceil(a * unit, b);
  const squaredLow = low * low / unit, squaredHigh = ceil(high * high, unit);
  let powerLow = low, powerHigh = high, sumLow = 0n, sumHigh = 0n, odd = 1n;
  for (;;) {
    sumLow += powerLow / odd; sumHigh += ceil(powerHigh, odd);
    powerLow = powerLow * squaredLow / unit;
    powerHigh = ceil(powerHigh * squaredHigh, unit); odd += 2n;
    if (powerHigh <= 2n) return [2n * sumLow, 2n * (sumHigh + ceil(9n * powerHigh, 8n * odd))];
  }
}

function lnTwo(precision: number): Interval {
  if (precision !== 160) return oddSeries(1n, 3n, precision);
  return defaultTwo ??= oddSeries(1n, 3n, precision);
}

export function logarithm(x: number, precision: number): Interval {
  const [m, e] = parts(x), width = bitLength(m), one = 1n << BigInt(width - 1);
  const k = BigInt(e + width - 1), [low, high] = oddSeries(m - one, m + one, precision);
  const [constantLow, constantHigh] = lnTwo(precision);
  return k >= 0n ? [low + k * constantLow, high + k * constantHigh]
    : [low + k * constantHigh, high + k * constantLow];
}

function lnTen(precision: number): Interval {
  if (precision !== 160) return logarithm(10, precision);
  return defaultTen ??= logarithm(10, precision);
}

function quotient(numerator: Interval, denominator: Interval, precision: number): Interval {
  const unit = 1n << BigInt(precision);
  let low: bigint | undefined, high: bigint | undefined;
  for (const n of numerator) for (const d of denominator) {
    const below = floor(n * unit, d), above = ceil(n * unit, d);
    if (low === undefined || below < low) low = below;
    if (high === undefined || above > high) high = above;
  }
  return [low as bigint, high as bigint];
}

function exponentialRange(low: bigint, high: bigint, positive: boolean, precision: number): readonly [bigint, bigint, number] | undefined {
  const unit = 1n << BigInt(precision);
  const [constantLow, constantHigh] = lnTwo(precision), k = low / constantHigh;
  const reducedLow = low - k * constantHigh, reducedHigh = high - k * constantLow;
  if (reducedLow < 0n || reducedHigh >= unit) return undefined;
  let termLow = unit, termHigh = unit, sumLow = unit, sumHigh = unit, n = 0n;
  for (;;) {
    n++;
    termLow = termLow * reducedLow / (unit * n);
    termHigh = ceil(termHigh * reducedHigh, unit * n);
    if (termHigh <= 1n) { sumHigh += 2n * termHigh; break; }
    sumLow += termLow; sumHigh += termHigh;
  }
  return positive ? [sumLow, sumHigh, Number(k) - precision]
    : [unit * unit / sumHigh, ceil(unit * unit, sumLow), -Number(k) - precision];
}

/** Enclose exp(integer / 2**precision) without a binary64 intermediate. */
export function exponentialEndpoint(integer: bigint, precision: number): readonly [number | null, number | null] | undefined {
  const cutoff = 1024n << BigInt(precision);
  if (integer >= cutoff) return [null, null];
  if (integer <= -cutoff) return [0, 0];
  if (integer === 0n) return [1, 1];
  const magnitude = integer < 0n ? -integer : integer;
  const range = exponentialRange(magnitude, magnitude, integer > 0n, precision);
  if (range === undefined) return undefined;
  const [low, high, exponent] = range;
  return [rounded(low, exponent), rounded(high, exponent)];
}

/** Internal kernel. The optional starting precision permits refinement tests. */
export function transcendental(kind: ElementaryKind, x: number, initialPrecision = 160): number | null {
  if (!Number.isFinite(x) || (kind !== 'exp' && x <= 0)) return null;
  if (kind === 'exp') {
    if (x >= 1024) return null;
    if (x <= -1024) return 0;
    if (Math.abs(x) < 2 ** -60) return 1;
  } else {
    if (x === 1) return 0;
    if (kind === 'log2') {
      const [m, e] = parts(x);
      if ((m & (m - 1n)) === 0n) return e + bitLength(m) - 1;
    }
    if (kind === 'log10') {
      const [m, e] = parts(x);
      let integer = e >= 0 ? m << BigInt(e)
        : m % (1n << BigInt(-e)) === 0n ? m >> BigInt(-e) : 0n;
      let count = 0;
      while (integer > 1n && integer % 10n === 0n) { integer /= 10n; count++; }
      if (integer === 1n) return count;
    }
  }
  for (let precision = Math.max(64, initialPrecision); ; precision += 80) {
    let low: bigint, high: bigint, exponent: number;
    if (kind === 'exp') {
      const [start, end] = fixed(x, precision);
      const range = exponentialRange(start, end, x >= 0, precision);
      if (range === undefined) continue;
      [low, high, exponent] = range;
    }
    else {
      const natural = logarithm(x, precision);
      [low, high] = kind === 'log' ? natural
        : quotient(natural, kind === 'log2' ? lnTwo(precision) : lnTen(precision), precision);
      exponent = -precision;
    }
    const left = rounded(low, exponent), right = rounded(high, exponent);
    if (Object.is(left, right)) return left;
  }
}
